platforms/core-execution/build-cache-local/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCache.java

```diff
@@ -146,9 +153,19 @@ public void storeLocally(HashCode key, File file) {
         });
     }
 
-    private void storeInsideLock(HashCode key, File file) {
-        LocallyAvailableResource resource = fileStore.move(key.toString(), file);
-        fileAccessTracker.markAccessed(resource.getFile());
+    private void storeInsideLock(HashCode key, File sourceFile) {
+        File targetFile = getCacheEntryFile(key);
+        try {
+            Files.move(sourceFile.toPath(), targetFile.toPath(), StandardCopyOption.ATOMIC_MOVE);
+        } catch (FileAlreadyExistsException ignore) {
+            // We already have the file in the build cache
+            // Note that according to the documentation of `Files.move()`, whether this exception is thrown
+            // is implementation specific: it can also happen that the target file gets overwritten, as if
+            // `REPLACE_EXISTING` was specified. This seems to match the behavior exhibited by `File.renameTo()`.
+        } catch (IOException e) {
+            throw new UncheckedIOException(String.format("Couldn't move cache entry '%s' into local cache: %s", key, e), e);
+        }
```

Comment:
💭 This is where we have switched from using `File.renameTo()` (which silently ignored not being able to move the file) to `Files.move()` which at least fails visibly.